export content of git submodules

pohly opened the pull request:
### What does this do / why do we need it?

The "git checkout-index" command is unaware of git submodules,
therefore their content was missed when exporting files. Only a few projects have submodules (for example, git-hooks in github.com/nightlyone/lockfile and idl in github.com/uber/jaeger-client-go) and it seems that so far those submodules were not needed.

But nonetheless, the extra files are part of the source tree and thus should be exported because some projects might depend on them.

### What should your reviewer look out for in this PR?

The tests for this feature depend on repos that I created under github.com/pohly. I can keep them there, but it might be cleaner to clone them into your own org and then update the tests.

pohly commented:
Please note that I should be covered by a company CLA between Intel and Google. Let me know if you need more details to find that CLA.

sdboyer commented:
Wow. Who'dve thought there'd be yet another reason i'd hate git submodules.

If you could provide a bit more info on the Intel<>Google CLA, that'd be helpful - being that i'm not a Google employee, i don't have ready access to that information, and i'd rather not trip over anything here.

> Does the usage of POSIX shell syntax work also on Mac OS and Windows?

On OSX, yes. i'm not sure about Windows. That tests pass on appveyor suggests it's _probably_ fine, but if you can think of any way to test this reasonably easily, it would be helpful.

Also, a CHANGELOG entry, please!

pohly commented:
sam boyer <notifications@github.com> writes:
> If you could provide a bit more info on the Intel<>Google CLA, that'd
>be helpful - being that i'm not a Google employee, i don't have ready
>access to that information, and i'd rather not trip over anything here.

I'll try to sort that out with some Google employees.

>> Does the usage of POSIX shell syntax work also on Mac OS and Windows?
>
> On OSX, yes. i'm not sure about Windows. That tests pass on appveyor
> suggests it's _probably_ fine,

I'm not sure whether existing tests cover this case. Is there any test
where a repo gets vendored which has submodules? Without those, "git
submodule foreach" will never invoke the shell command and might pass
on Windows because of that.

I had a brief look at how tests are done, but it looked a bit daunting,
so I wanted to send this PR first to get some opinions before investing
more time.

> but if you can think of any way to test
> this reasonably easily, it would be helpful.

Construct two repos, include the first one in the second one with a
submodule, then vendor the second one into a project. Check that the
resulting "vendor" directory contains the files from the first repo.

> Also, a CHANGELOG entry, please!

I had already added one in commit 078c630. The way I understand it, one
has to create a PR to get a pull request number, then force-push a
commit with changelog entry - at least that's how I did it.

pohly commented:
@sdboyer I've rebased and added a testcase with recursive submodules. I don't think there were any test cases which had a submodule at all. The new test revealed that my original code only worked for exporting submodules from the main repo - it failed for submodules inside submodules.

In order to test this, I had to extend the integration testing a bit. Otherwise it would not have detected the missing submodule files. The new testing should now confirm whether the code also works on non-Linux platforms.

Regarding the CLA: I'm still working on it :-/ In the meantime consider this a RFC path, and *not a contribution* - it cannot be merged yet.

pohly commented:
About the test: the new repos are under github.com/pohly and I can keep them there in perpetuity, but it might be better to move them to github.com/sdboyer so that they are under your control. You don't need a CLA for that, do you?

pohly commented:
Hmm, it failed on appvoyer once (https://ci.appveyor.com/project/golang/dep/build/3389) but after I changed the appvoyer config to be more verbose it succeeded without any other changes.

sdboyer commented:
Cool - i'll aim to have a look at this next week, as this week is focused on shipping v0.5.0.

pohly commented:
I signed it!

This is for the CLA bot - I should be on the approved contributor list now.

pohly commented:
@sdboyer I understand that your priorities may have shifted now, but as I depend on "dep" and go modules don't look like a viable alternative it would be nice to get this merged.

sdboyer commented on gps/vcs_source.go:
```diff
@@ -203,6 +203,34 @@ func (s *gitSource) exportRevisionTo(ctx context.Context, rev Revision, to strin
+			// "<root>/foo" for bar. By stripping our root
+			// directory we get a relative path that we can
+			// append to the target directory.
+			"(cd $(grep gitdir .git | sed -e 's;gitdir: *;;' .git) && mv index origindex && trap 'mv origindex index' EXIT && git read-tree \"$sha1\" && git checkout-index -a --prefix=\""+to+"$(echo \"$toplevel\" | sed -e 's;"+r.LocalPath()+";;')/$path/\")",
```
`grep`, `sed`, etc., - not gonna be portable to windows 😢

Let's take an approach with two execs - call out to git to enumerate the submodules, loop over the results with a call to `fs.RenameWithFallback()`, then call out again with `git submodule foreach`. Simplicity, correctness and portability are worth the performance hit.

sdboyer replied:
i think you should be able to get what you need from `git submodule status` output. Icky, but it's there.

pohly replied:
FWIW, you mentioned before (https://github.com/golang/dep/pull/1909#issuecomment-402057799) that the shell code might be okay because "tests pass on appveyor suggests it's probably fine". That was before I added an explicit test for submodules, but it continued to pass also after I added the new testcase. So just to be sure, automatic testing in appveyor includes running dep on Windows?

But I've rewritten the code such that it is based on the output of `git submodule status` anyway, with the "export tree" functionality done by the same Go code as before, just moved to a new method. It's easier to understand this way. The only thing I don't like about it is parsing the output of `git submodule status`, because there is no well-defined, machine-readable output format.

sdboyer commented on internal/fs/fs.go:
```diff
@@ -692,3 +692,50 @@ func volumeName(path string) (v string) {
+
+// CopyTree recursively copies a source directory into a target directory.
+// Target directories are created if they don't exist yet.
+func CopyTree(dest, src string) error {
```
We didn't have something like this already?

sdboyer replied:
We've got `CopyDir` - https://github.com/golang/dep/blob/22125cfaa6ddc71e145b1535d4b7ee9744fefff2/internal/fs/fs.go#L356

i think that should do it.

pohly replied:
I've added a separate commit which removes this code duplication.

This PR is starting to collect some commits which could also be merged separately, like that cleanup commit. Please let me know whether I should create separate PRs for that or whether you are fine with merging everything in one go.

mvdan commented:
Dep was officially deprecated earlier this year, and the proposal to archive this repository [was accepted](https://github.com/golang/go/issues/38158). As such, I'm closing outstanding issues before archiving the repository. For any further comments, please use the proposal thread on the Go issue tracker. Thanks!
